### DISCUSSION TITLE
SSL_CTX_set_client_hello_cb resume after SSL_CLIENT_HELLO_RETRY

### DISCUSSION THREAD

**Discussion by mgdaniel** [openssl/openssl#25513] [Q&A]
Am using a hello callback where mostly it returns SSL_CLIENT_HELLO_SUCCESS but occasionally returns SSL_CLIENT_HELLO_RETRY after kicking off some other-thread intermediate processing.  How do I resume processing of the handshake after the intermediate processing concludes?  TIA.

**Comment by mgdaniel** [comment]
Seems a silly question now I have (re)discovered (after a number of years) SSL_do_handshake(). TAW.

**Reply by vyasatavi** [reply]
On this topic @mattcaswell it would be nice to have the Client hello CB after s->hit is evaluated. This is specifically for HSM cases where we often rely on a separate thread to process connections that require offload to a network HSM and the main thread is not affected. Some engines don't support Async so this dual-thread mechanism becomes necessary.

**Reply by mattcaswell** [reply]
Seems like this should be a separate issue. But the whole design of the client hello cb is that it is called very early and *before* extensions are processed - which necessarily means this is before `s->hit` is set.
